### PR TITLE
Fixes https://github.com/tpm2-software/tpm2-pkcs11/issues/26

### PR DESCRIPTION
**Pull request by robby-cornelissen**
This should fix the aforementioned issue by copying the `TPM2_PKCS11_TCTI` environment variable so that any modifications made don't impact the original value.

Maybe I've overstepped by doing a little bit of refactoring (clearer variable names and reduced conditional branches), but I leave that to the discretion of the maintainer.

Disclaimer: I'm interested in this project and would like to contribute further, but have never written a line of C before. I tried to do my research and due diligence as best as possible.

### CODE REVIEW

File: src/lib/tcti_ldr.c [hunk 1]
```diff
@@ -168,49 +168,40 @@ tcti_conf tcti_get_config(void) {
     };
 
     /* no tcti config supplied, get it from env */
-    char *optstr = getenv (TPM2_PKCS11_TCTI);
-    if (!optstr) {
+    char *env_conf_string = getenv(TPM2_PKCS11_TCTI);
```
Comment (williamcroberts) [inline src/lib/tcti_ldr.c]: Their is a bug, but your patch doesn't do the bare minimum to fix that bug. Ie 1 patch/commit to fix the bug and another if you want to do refactoring.

I don't see optstr and env_conf_string needing to exist.... you can just "reseat the pointer on optstr" when you need to do the duplication below.

File: src/lib/tcti_ldr.c [hunk 2]
```diff
@@ -168,49 +168,40 @@ tcti_conf tcti_get_config(void) {
...
+    if (!env_conf_string) {
         /* nothing user supplied, use default */
         return conf;
     }
 
-    char *split = strchr(optstr, ':');
+    char *conf_string = malloc(strlen(env_conf_string) + 1);
+    strcpy(conf_string, env_conf_string);
```
Comment (williamcroberts) [inline src/lib/tcti_ldr.c]: use `strdup()` versus malloc and strcpy

File: src/lib/tcti_ldr.c [hunk 3]
```diff
@@ -168,49 +168,40 @@ tcti_conf tcti_get_config(void) {
...
+         */
+        split[0] = '\0';
 
-    /* Case B */
-    if (!optstr[0]) {
-        conf.opts = &split[1];
-        return conf;
-    }
+        /* Case C, D */
+        if (conf_string[0]) {
```
Comment (williamcroberts) [inline src/lib/tcti_ldr.c]: You're stated goal here was to reduce branches, but you added a branch and indentation under an else clause to get to the point of a signgle return. This could be all done with else if's but this refactor should be in a different patch.

File: src/lib/tcti_ldr.c (hunk 2 above)
Comment (williamcroberts) [reply src/lib/tcti_ldr.c]: You also need to free this. Below you try and get to a single return point, so you'll need a free before the return.

**Comment by williamcroberts** [comment]
FYI That test that introduces this bug will be getting deleted.

File: src/lib/tcti_ldr.c [hunk 4]
```diff
@@ -168,49 +168,40 @@ tcti_conf tcti_get_config(void) {
...
+            conf.name = fixup_name(conf_string);
```
Comment (williamcroberts) [inline src/lib/tcti_ldr.c]: additionally, fixup name can return static data which can't be free'd. Thus you would need to heap allocate everything.

File: src/lib/tcti_ldr.c (hunk 1 above)
Comment (robby-cornelissen) [reply src/lib/tcti_ldr.c]: Thanks for your feedback. I'm definitely not as nimble in C as I am in other programming languages, and I appreciate you taking the time to help me along. I'll try to explain the thought process that went into this patch a little further, so that any flaws in my reasoning become more apparent.

1. Your point about not including unnecessary refactoring in a bug fix is well taken. I didn't introduce the refactoring on a whim, but was trying to address two issues:
    - Branching: I indeed expected to have to `free()` the `conf_string` before the method return (more on that later). Having to do that would have introduced a fairly inelegant label with a number of corresponding `goto` statements, so, instead of doing that, I decided to refactor the branches.
    - Variable naming: The method is trying to populate a `tcti_conf` string with `name` and `opts` fields. Since the `optstr` variable is expected to contain information about both fields, it felt like `confstr` would be a more descriptive name. I then went from `confstr` to `conf_string` based on the coding standards document included in this project (multiple words separated by underscore, no abbreviations except for some exceptions). Rigorously applying the coding standard would have led to `configuration_string` and `environment_configuration_string`, but this would have been inconsistent with the rest of the project (`tcti_conf`, `getenv()` etc.), so I decided against it.
2. `optstr` and `env_conf_string` indeed don't need to coexist, and my patch introduced one as a replacement for the other. As also touched upon above, the renaming came about because I was unsure what name I would use for the copied `optstr`. Both `tmp_optstr` and `optstr_copy` felt a little yuck.
3. I decided on a blanket copy of the environment variable because of [uncertainty](https://stackoverflow.com/questions/19745648/when-getenv-can-alter-content-of-previously-returned-pointer-when-using-glibc) about how subsequent calls to `setenv()` or `putenv()` could potentially affect the value that the pointer obtained from `getenv()` is pointing to.

File: src/lib/tcti_ldr.c (hunk 2 above)
Comment (robby-cornelissen) [reply src/lib/tcti_ldr.c]: 1. Agree on using `strdup()` and checking whether memory was successfully allocated.
2. Here is where I'm a little confused and might lack some basic understanding. As noted in my previous comment, it was my intention to free the memory, but calling `free(conf_string)` before returning made all tests fail. My reasoning was that in my implementation, since both fields of `tcti_conf` point to a location in `conf_string`, freeing `conf_string` results in `tcti_conf` no longer having access to its data. This seemed to be confirmed by my tests. Surely I'm misunderstanding something, but I don't quite know what.

File: src/lib/tcti_ldr.c (hunk 3 above)
Comment (robby-cornelissen) [reply src/lib/tcti_ldr.c]: Agree on the general principle of separating bug fixes from refactoring.

**Comment by williamcroberts** [comment]
https://github.com/tpm2-software/tpm2-pkcs11/pull/29
